### PR TITLE
Fixed segmentation issue that mostly appeared on DWI data

### LINKED ISSUE #2464: ValueError: min() arg is an empty sequence caused by empty segmentation
For some reasons, a DWI dataset yields an empty segmentation, which causes the postprocessing module to crash. 

We should:
- figure out why the segmentation is failing
- raise a more elegant error during postprocessing if inputing an empty segmentation

To reproduce:

Data: 
[sub-brno02_dwi_crop_moco_dwi_mean.nii.gz](https://github.com/neuropoly/spinalcordtoolbox/files/3672918/sub-brno02_dwi_crop_moco_dwi_mean.nii.gz)
[sub-perform02_dwi_dwi_mean.nii.gz](https://github.com/neuropoly/spinalcordtoolbox/files/3673498/sub-perform02_dwi_dwi_mean.nii.gz)

SCT commit: 2b85953c2f22b0d9225a148173631313b8c3566d
Syntax (with error):
~~~
sct_deepseg_sc -i sub-brno02_dwi_crop_moco_dwi_mean.nii.gz -c dwi -r 0

--
Spinal Cord Toolbox (git-jca/2458-crop-2b85953c2f22b0d9225a148173631313b8c3566d)


Method:
	Centerline algorithm: svm
	Assumes brain section included in the image: False
	Dimension of the segmentation kernel convolutions: 2d


Create temporary folder (/var/folders/s8/4qnm5q1n261ch35b5kkclsb00000gn/T/sct-20190930152748.787127-p3a3u7kb)...
Reorient the image to RPI, if necessary...
Finding the spinal cord centerline...
Remove temporary files...
rm -rf /var/folders/s8/4qnm5q1n261ch35b5kkclsb00000gn/T/sct-20190930152748.955878-z_k4czod
Resample the image to 0.5x0.5 mm in-plane resolution...
load data...
load data...
Cropping the image around the spinal cord...
Normalizing the intensity...
Segmenting the spinal cord using deep learning on 2D patches...
Reassembling the image...
WARNING: To avoid intensity overflow due to convertion to uint8, intensity will be rescaled to the maximum quantization scale.
Resampling the segmentation to the native image resolution using linear interpolation...
Binarizing the resampled segmentation...
Traceback (most recent call last):
  File "/Users/julien/code/sct/scripts/sct_deepseg_sc.py", line 208, in <module>
    main()
  File "/Users/julien/code/sct/scripts/sct_deepseg_sc.py", line 178, in main
    remove_temp_files=remove_temp_files, verbose=verbose)
  File "/Users/julien/code/sct/spinalcordtoolbox/deepseg_sc/core.py", line 578, in deep_segmentation_spinalcord
    im_seg_r_postproc = post_processing_volume_wise(im_seg_r)
  File "/Users/julien/code/sct/spinalcordtoolbox/deepseg_sc/postprocessing.py", line 139, in post_processing_volume_wise
    im_seg = _remove_isolated_voxels_on_the_edge(im_seg)
  File "/Users/julien/code/sct/spinalcordtoolbox/deepseg_sc/postprocessing.py", line 110, in _remove_isolated_voxels_on_the_edge
    metrics, _ = compute_shape(im_seg, angle_correction=False)
  File "/Users/julien/code/sct/spinalcordtoolbox/process_seg.py", line 61, in compute_shape
    min_z_index, max_z_index = min(Z), max(Z)
ValueError: min() arg is an empty sequence
~~~

### PR DESCRIPTION
**Pull request by jcohenadad**
For some reasons, some DWI datasets yield empty segmentations (or highly sub-segmented), causing the postprocessing module to crash later on.

This PR fixes:
- The segmentation issue, which was related to a resampling problem.
- Raises a more elegant error during postprocessing if inputing an empty segmentation.

Fixes #2464 

**WARNING: Output segmentation is overall larger (especially on dwi scans), therefore CSA studies should be based on this SCT version (or previous version), but not a mix of both versions.**

### CODE REVIEW

File: spinalcordtoolbox/deepseg_sc/core.py [hunk 1]
```diff
@@ -111,38 +111,28 @@ def find_centerline(algo, image_fname, contrast_type, brain_bool, folder_output,
...
-        logger.info("Resample the image to 0.5x0.5 mm in-plane resolution...")
-        fname_res = sct.add_suffix(image_fname, '_resampled')
-        input_resolution = Image(image_fname).dim[4:7]
-        new_resolution = 'x'.join(['0.5', '0.5', str(input_resolution[2])])
-
-        resampling.resample_file(image_fname, fname_res, new_resolution, 'mm', 'linear', verbose=0)
-
-        resampling.resample_file(centerline_filename, centerline_filename, new_resolution, 'mm', 'linear', verbose=0)
+    # TODO: for some reason, when algo == 'file', the absolutepath is changed to None out of the method find_centerline
+    im_ctl.absolutepath = ctl_absolute_path
```
Comment (charleygros) [inline spinalcordtoolbox/deepseg_sc/core.py]: For some reason, when algo == 'file', I could not set the absolute path of `im_ctl`, while it works with `svm` and `viewer`.
@jcohenadad : Any ideas?

Comment (jcohenadad) [reply spinalcordtoolbox/deepseg_sc/core.py]: no quick idea unfortunately. I will have to run the debugger to understand the cause of the problem.

**Comment by charleygros** [comment]
Centerline on `dmri` data is now correct, but segmentation fails.
Seems all good for the other contrasts.

Try to compare between `master` and current branch:
- `im_crop_nii`: similar.
- `im_norm_in`: slightly different, but really small changes: I would be surprised if that's the reason why the segmentation fails.

File: spinalcordtoolbox/deepseg_sc/core.py (hunk 1 above)
Comment (jcohenadad) [reply spinalcordtoolbox/deepseg_sc/core.py]: btw: an alternative would be to specify file name upon saving:
~~~
im_ctl.save(ctl_absolute_path)
~~~

**Comment by jcohenadad** [comment]
> Centerline on `dmri` data is now correct, but segmentation fails.
> Seems all good for the other contrasts.
> 
> Try to compare between `master` and current branch:
> 
> * `im_crop_nii`: similar.
> * `im_norm_in`: slightly different, but really small changes: not sure if that's the reason why the segmentation fails.

Sorry I'm not sure I fully understand what was done here. Did you compare the output segmentation from the master branch, with the segmentation run on this branch using either `im_crop_nii` (similar than master branch?) or `im_norm_in` (slightly different than master branch?) as an input of `segment_2d()`?

**Comment by charleygros** [comment]
> Sorry I'm not sure I fully understand what was done here. Did you compare the output segmentation from the master branch, with the segmentation run on this branch using either `im_crop_nii` (similar than master branch?) or `im_norm_in` (slightly different than master branch?) as an input of `segment_2d()`?

Sorry, it was not very clear indeed.
Update: segmentation fails on dmri images on both `master` and current branch.
As the temp files (eg `im_norm_nii`) are similar between the two branches: I assume that the segmentation issue does not come from the recent changes (ie current branch).

**Comment by charleygros** [comment]
Segmentation failure on `dmri` data observed with SCT 3.2.7.

**Comment by charleygros** [comment]
> Update: segmentation fails on dmri images on both `master` and current branch.

Nevermind --> Was testing on an inappropriate data. Did some more extensive testing: it seems to be fixed on my side. Sorry about that.

@jcohenadad: Could you please test on your data as well? :)

**Comment by jcohenadad** [comment]
getting this error (syntax + commit number included):
~~~
sct_deepseg_sc -i /Users/julien/data/spineGeneric_20190925/results/data/sub-perform02/dwi/sub-perform02_dwi_crop_moco_dwi_mean_crop.nii.gz -c dwi -r 0 -v 2
...
Saving image to /Users/julien/data/spineGeneric_20190925/results/data/sub-perform02/dwi/sub-perform02_dwi_crop_moco_dwi_mean_crop_seg.nii.gz orientation RPI shape (25, 29, 15)
Traceback (most recent call last):
  File "/Users/julien/code/sct/scripts/sct_deepseg_sc.py", line 208, in <module>
    main()
  File "/Users/julien/code/sct/scripts/sct_deepseg_sc.py", line 198, in main
    im_ctr.save(fname_ctr)
AttributeError: 'NoneType' object has no attribute 'save'
~~~

**Comment by jcohenadad** [comment]
> > Update: segmentation fails on dmri images on both `master` and current branch.
> 
> Nevermind --> Was testing on an inappropriate data. Did some more extensive testing: it seems to be fixed on my side. Sorry about that.
> 
> @jcohenadad: Could you please test on your data as well? :)

hum... doesn't seem to be fixed. I get the results below using commit f4019b2ccfeda946eaad5118f90878a8a581db4e on [this image](https://github.com/neuropoly/spinalcordtoolbox/files/3672918/sub-brno02_dwi_crop_moco_dwi_mean.nii.gz):

anim

I am wondering if the lack of performance comes from the normalization, because the background/cord contrast is already pretty low, and the edge of the image contains null voxels, meaning that the normalization procedure, which take all range of values in the image, flattens the cord/background contrast. 

Doing some investigations by cropping the image and running the segmentation again produces very different results, suggesting there might be something going on with the normalization

**Comment by jcohenadad** [comment]
after removing the "null" edges, the segmentation is still suboptimal:
~~~
sct_crop_image -i sub-brno02_dwi_crop_moco_dwi_mean.nii.gz -xmin 5 -xmax -5 -ymin 5 -ymax -5
sct_deepseg_sc -i sub-brno02_dwi_crop_moco_dwi_mean_crop.nii.gz -c dwi
~~~
anim

**Comment by charleygros** [comment]
The normalisation step is quite critical indeed.
Could you please point me to where your image is saved? Or copy it under `duke/sct_testing/issues/`?
Thanks

**Comment by jcohenadad** [comment]
> The normalisation step is quite critical indeed.
> Could you please point me to where your image is saved? Or copy it under `duke/sct_testing/issues/`?
> Thanks

The image is in the issue associated with this PR

**Comment by charleygros** [comment]
> getting this error (syntax + commit number included):
> 
> ```
> sct_deepseg_sc -i /Users/julien/data/spineGeneric_20190925/results/data/sub-perform02/dwi/sub-perform02_dwi_crop_moco_dwi_mean_crop.nii.gz -c dwi -r 0 -v 2
> ...
> Saving image to /Users/julien/data/spineGeneric_20190925/results/data/sub-perform02/dwi/sub-perform02_dwi_crop_moco_dwi_mean_crop_seg.nii.gz orientation RPI shape (25, 29, 15)
> Traceback (most recent call last):
>   File "/Users/julien/code/sct/scripts/sct_deepseg_sc.py", line 208, in <module>
>     main()
>   File "/Users/julien/code/sct/scripts/sct_deepseg_sc.py", line 198, in main
>     im_ctr.save(fname_ctr)
> AttributeError: 'NoneType' object has no attribute 'save'
> ```

Regarding this error: This is due to `-v 2` --> This commented lines: [here](https://github.com/neuropoly/spinalcordtoolbox/blob/master/spinalcordtoolbox/deepseg_sc/core.py#L562).

Happy to fix this. Quick question: What is the best way to setup a debug mode?
What about creating a `DEBUG` variable at the top of the script such as: [here](https://github.com/neuropoly/spinalcordtoolbox/blob/fc520c3ccd5cf0cd629ea70d134221b57eb9db68/install/create_package.py#L11).

**Comment by jcohenadad** [comment]
> > getting this error (syntax + commit number included):
> > ```
> > sct_deepseg_sc -i /Users/julien/data/spineGeneric_20190925/results/data/sub-perform02/dwi/sub-perform02_dwi_crop_moco_dwi_mean_crop.nii.gz -c dwi -r 0 -v 2
> > ...
> > Saving image to /Users/julien/data/spineGeneric_20190925/results/data/sub-perform02/dwi/sub-perform02_dwi_crop_moco_dwi_mean_crop_seg.nii.gz orientation RPI shape (25, 29, 15)
> > Traceback (most recent call last):
> >   File "/Users/julien/code/sct/scripts/sct_deepseg_sc.py", line 208, in <module>
> >     main()
> >   File "/Users/julien/code/sct/scripts/sct_deepseg_sc.py", line 198, in main
> >     im_ctr.save(fname_ctr)
> > AttributeError: 'NoneType' object has no attribute 'save'
> > ```
> 
> Regarding this error: This is due to `-v 2` --> This commented lines: [here](https://github.com/neuropoly/spinalcordtoolbox/blob/master/spinalcordtoolbox/deepseg_sc/core.py#L562).
> 
> Happy to fix this. Quick question: What is the best appropriate way to setup a debug mode in the code?
> What about creating a `DEBUG` variable at the top of the script such as: [here](https://github.com/neuropoly/spinalcordtoolbox/blob/fc520c3ccd5cf0cd629ea70d134221b57eb9db68/install/create_package.py#L11).

Hum. I would uncomment the line (but not the todo). I would like to minimize the use of a global DEBUG. It happened in the past that we would forget to set these back to 0. Ideally we should have a config file, external from SCT repos, where we would set this variables.

**Comment by charleygros** [comment]
> Hum. I would uncomment the line (but not the todo). I would like to minimize the use of a global DEBUG. It happened in the past that we would forget to set these back to 0. Ideally we should have a config file, external from SCT repos, where we would set this variable.

Done! And same for the viewer labels.

However, after the resampling (back to initial resolution), we lose the center point on few slices + not binarised (since we use interpolation = linear).
We could dilate/resample/center_of_mass. But was wondering if we have something already implemented under centerline or labels_utils that deals with this (ie resampling of centerline / labels)?

**Comment by charleygros** [comment]
@jcohenadad: This is not the best solution but... I got a good segmentation on both images by setting this [threshold](https://github.com/neuropoly/spinalcordtoolbox/blob/master/spinalcordtoolbox/deepseg_sc/core.py#L395) to 0.

Extensive testing on `dmri` data could be of interest.

On a side note, we could also think of adding this threshold to the sct_function parameters.

**Comment by jcohenadad** [comment]
> However, after the resampling (back to initial resolution), we lose the center point on few slices + not binarised (since we use interpolation = linear).

@ you are only referring to the "-v 2" output anyway, right? so this has no impact on regular use of the function. If that's indeed the case, then do we really need to resample to the native resolution? I mean: do we need this output for debugging? (we could assess the quality of the centerline in the 0.5mm resolution, right?)

**Comment by jcohenadad** [comment]
> @jcohenadad: This is not the best solution but... I got a good segmentation on both images by setting this [threshold](https://github.com/neuropoly/spinalcordtoolbox/blob/master/spinalcordtoolbox/deepseg_sc/core.py#L395) to 0.
> 
> Extensive testing on `dmri` data could be of interest.
> 
> On a side note, we could also think of adding this threshold to the sct_function parameters.

@charleygros great catch! I will run deepseg_sc with threshold=0 on all contrasts. Based on the results, we will decide what to do (set a specific threshold for dwi? same for all contrasts? what threshold value?)

**Comment by charleygros** [comment]
> > However, after the resampling (back to initial resolution), we lose the center point on few slices + not binarised (since we use interpolation = linear).
> 
> @charleygros you are only referring to the "-v 2" output anyway, right? so this has no impact on regular use of the function. If that's indeed the case, then do we really need to resample to the native resolution? I mean: do we need this output for debugging? (we could assess the quality of the centerline in the 0.5mm resolution, right?)

Yes that's right: I was referring to the "-v 2", no impact for the segmentation, makes more sense to assess the centerline in the 0.5x0.5mm2 resolution --> we could simply save this centerline into the temp folder and use this one for debugging.
--> Will do!

**Comment by jcohenadad** [comment]
OK, after running `sct_deepseg_sc` on a few subjects and contrasts, I confirm that the segmentation is generally better for threshold=0, compared to threshold=0.5.

In some instances however, the segmentation was missing with threshold=0 (for reasons that need further investigations), like for the example below:

threshold=0
Screen Shot 2019-10-03 at 2 21 07 PM

threshold=0.5
Screen Shot 2019-10-03 at 2 21 16 PM

**Comment by charleygros** [comment]
> OK, after running `sct_deepseg_sc` on a few subjects and contrasts, I confirm that the segmentation is generally better for threshold=0, compared to threshold=0.5.
When you said "few ... contrasts," --> did you mean that it was better on T2w, T1w ...etc. as well?


> In some instances however, the segmentation was missing with threshold=0 (for reasons that need further investigations), like for the example below:
Interesting! Happy to look into it. Is the data on `sct_testing`?

**Comment by jcohenadad** [comment]
Yes, better for t2, T1w, etc

Yes it would be nice if you could look into it :-)

I used data  from spine generic: amu02, ucl01, beijingGE01

**Comment by lrouhier** [comment]
Hi, it seems to be better for my test image. I have something similar to what Julien have in his previous comment (last slice missing). I will run it on all images and will let you know.
image

**Comment by jcohenadad** [comment]
> Hi, it seems to be better for my test image. I have something similar to what Julien have in his previous comment (last slice missing). I will run it on all images and will let you know.
> image

thanks for the feedback @lrouhier -- in the future can you please indicate SCT commit + data used. Also when doing a comparison you need to show the before/after (otherwise no comparison possible from our perspective). Thx

**Comment by lrouhier** [comment]
*sct_version* : git-jca/2464-deepseg_sc-empty-ac5357e3a73253ce6877d738818e2751f248839e
*path_to_image*:/duke/projects/ml_sci_prognosis/issue/issue_deepseg_#2470
*command run*: sct_deepseg_sc -i axial_t2_ANU-004.nii -brain 0 -qc

**Comment by jcohenadad** [comment]
> _command run_: sct_deepseg_sc -i axial_t2_ANU-004.nii -brain 0 -qc

this is the wrong syntax (-c parameter is missing). If you copy/paste the output of the qc like i did [here](https://github.com/neuropoly/spinalcordtoolbox/pull/2467#issuecomment-538067061), it (i) prevents you from making copy/paste errors and (ii) enables easier before/after comparison (same FOV and cropping-- easier to compare images).

**Comment by lrouhier** [comment]
It seems to work better on the whole dataset. 
*Master version*
image

*modified version*
image

*image in*:/Volumes/projects/ml_sci_prognosis/issue/issue_deepseg_#2470

**Comment by baxpr** [comment]
For what it's worth - with the 4.0.2 release I was getting an empty segmentation from sct_deepseg_sc on a more standard T2W image (not DWI) with 0.3x0.3x5.0 mm voxels that used to segment fine under an earlier version 4 (didn't record which). This commit seems to have fixed that and it's working again.

**Comment by jcohenadad** [comment]
Awesome, thanks for the feedback @lrouhier and @baxpr. We will make a few additional investigations before merging this PR, notably fix the problem of the edge slices, and look at the stability of the algorithm across resolutions and contrasts. This might take a few more days.

**Comment by jcohenadad** [comment]
The empty slice at z=0 is caused by `postprocessing._remove_isolated_voxels_on_the_edge()`, where the output of `metrics, _ = compute_shape(im_seg, angle_correction=False)` includes a nan:
```python
metrics['area'].data
array([      nan, 66.609375, 69.671875, 59.71875 , 67.375   , 68.140625,
       61.25    , 64.3125  , 65.84375 , 60.484375, 71.203125, 65.078125,
       62.78125 , 65.84375 , 65.84375 ])
```

**Comment by jcohenadad** [comment]
OK the missing segmentation at the edge has been fixed in https://github.com/neuropoly/spinalcordtoolbox/pull/2467/commits/f6c71d1560f9988ed3794130f42f2e9a102ac6f2 (stupid mistake: forgot to divide the median by 2 to find outlier CSA at the edge)

**Comment by jcohenadad** [comment]
Results across multiple contrasts, subjects, vendors and resolution:
https://www.dropbox.com/s/60aawmnzii3l4gv/20191005_SCT_issue2464_qc.zip?dl=0

Conclusion: Works great, except for the MToff contrast, but I don't think this is related to the issue related to this PR.

**Comment by jcohenadad** [comment]
Results of batch_processing on OSX:

branch jca/2464-deepseg_sc-empty (commit c801fcd85368fdc5d27cdda7059429d7a2dbfc78):
~~~
t2/CSA:          74.49902509135494
mt/MTR(WM):      54.22596560112912
t2s/CSA_GM:      12.487834828856178
t2s/CSA_WM:      66.44240519216386
dmri/FA(CST_r):  0.7890679073313994
dmri/FA(CST_l):  0.7609108835133579
~~~

master 
~~~
t2/CSA:          73.98363895856627
mt/MTR(WM):      54.29594350190386
t2s/CSA_GM:      12.487834828856178
t2s/CSA_WM:      64.97073246634238
dmri/FA(CST_r):  0.7934020965216604
dmri/FA(CST_l):  0.7673452412911614
~~~

**Pull request merged**